### PR TITLE
tests: Improve coverage on signature v4 tests.

### LINKED ISSUE #3065: [tests] Add missing cases for signature-v4.go.
Add missing cases for `doesPresignedSignatureMatch` and `doesSignatureMatch` from signature-v4-.go

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Improves coverage across signature v4 tests.

## Motivation and Context
On going improvement in coverage and new tests.

Fixes #3065

## How Has This Been Tested?
Manually using `go test`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/signature-v4-parser.go [hunk 1]
```diff
@@ -83,11 +83,14 @@ func parseSignature(signElement string) (string, APIErrorCode) {
...
+	if signFields[1] == "" {
+		return "", ErrMissingFields
+	}
 	signature := signFields[1]
 	return signature, ErrNone
 }
 
-// Parse signed headers string.
+// Parse slice of signed headers from signed headers tag.
 func parseSignedHeaders(signedHdrElement string) ([]string, APIErrorCode) {
```
Comment (fwessels) [inline cmd/signature-v4-parser.go]: Maybe parseSignedHeader (no 's') would be a more appropriate name, as we are only processing a single element of the header one at a time.

File: cmd/signature-v4_test.go [hunk 2]
```diff
@@ -221,6 +227,63 @@ func TestDoesPresignedSignatureMatch(t *testing.T) {
...
+		// (8) Should error if the request is not ready yet, ie X-Amz-Date is in the future.
+		{
+			queryParams: map[string]string{
+				"X-Amz-Algorithm":      signV4Algorithm,
+				"X-Amz-Date":           now.Add(1 * time.Hour).Format(iso8601Format),
+				"X-Amz-Expires":        "60",
+				"X-Amz-Signature":      "badsignature",
+				"X-Amz-SignedHeaders":  "host;x-amz-content-sha256;x-amz-date",
+				"X-Amz-Credential":     fmt.Sprintf(credentialTemplate, serverConfig.GetCredential().AccessKeyID, now.Format(yyyymmdd), serverConfig.GetRegion()),
+				"X-Amz-Content-Sha256": payload,
```
Comment (krishnasrinivas) [inline cmd/signature-v4_test.go]: "X-Amz-Content-Sha256" can be removed as it's not part of the query params list http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html

It's not harmful to have it, just that the test case suggests that it is also required. you can put a comment if you want to have it.

(same with other test cases)

File: cmd/signature-v4-parser.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/signature-v4-parser.go]: :+1:

**Comment by fwessels** [comment]
LGTM

File: cmd/signature-v4_test.go [hunk 3]
```diff
@@ -100,6 +100,12 @@ func TestDoesPolicySignatureMatch(t *testing.T) {
...
 
 func TestDoesPresignedSignatureMatch(t *testing.T) {
+	rootPath, err := newTestConfig("us-east-1")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer removeAll(rootPath)
+
 	// sha256 hash of "payload"
 	payload := "239f59ed55e737c77147cf55ad0c1b030b6d7ee748a7426952f9b852d5a935e5"
```
Comment (krishnasrinivas) [inline cmd/signature-v4_test.go]: payload can be renamed to payloadSHA256 (not part of this patch, but just happened to notice)

**Comment by krishnasrinivas** [comment]
LG other than two minor comments (one unrelated to this patch)

File: cmd/signature-v4_test.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/signature-v4_test.go]: X-Amz-Content-Sha256 can be set by some aws-sdks already to do server validation and it is supported. 

So we don't need to remove it - s3 docs are outdated :smile:

Comment (harshavardhana) [reply cmd/signature-v4_test.go]:
```
 var AWS = require('aws-sdk');

 var s3 = new AWS.S3({computeChecksums: true, signatureVersion: 'v4'}); // this is the default setting                                            
 var params = {Bucket: 'myBucket', Key: 'myKey', Body: 'EXPECTED CONTENTS11'};
 var url = s3.getSignedUrl('putObject', params);
 console.log("The URL is", url);

```

```
$ node test.js 
The URL is https://s3.amazonaws.com/myBucket/myKey?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=4932f102c3944718186c8bcc4d1265f33269213d910f51264b6d165f945fc6ed&X-Amz-Credential=USWUXHGYZQYFYFFIT3RE%2F20161106%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161106T193317Z&X-Amz-Expires=900&X-Amz-Signature=0e4c884c54f7e6701eb5b755b0756fe6c656656d3dcd2c547633bdbcc18ff36c&X-Amz-SignedHeaders=host%3Bx-amz-content-sha256
```
